src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -1653,65 +1641,64 @@ abstract class RefChecks extends Transform {
       }
     }
 
-    private def isSimpleCaseApply(tree: Tree): Boolean = {
-      val sym = tree.symbol
-      def isClassTypeAccessible(tree: Tree): Boolean = tree match {
-        case TypeApply(fun, targs) =>
-          isClassTypeAccessible(fun)
-        case Select(module, apply) =>
-          // scala/bug#4859 `CaseClass1().InnerCaseClass2()` must not be rewritten to `new InnerCaseClass2()`;
-          //          {expr; Outer}.Inner() must not be rewritten to `new Outer.Inner()`.
-          treeInfo.isQualifierSafeToElide(module) &&
+    private def isSimpleCaseApply(fun: Tree): Boolean = {
+      val sym = fun.symbol
+      def isClassTypeAccessible = {
+        val Select(module, _) = fun: @unchecked
+        // scala/bug#4859 `CaseClass1().InnerCaseClass2()` must not be rewritten to `new InnerCaseClass2()`;
+        //          {expr; Outer}.Inner() must not be rewritten to `new Outer.Inner()`.
+        treeInfo.isQualifierSafeToElide(module) &&
           // scala/bug#5626 Classes in refinement types cannot be constructed with `new`.
           !module.exists { case t @ Select(_, _) => t.symbol != null && t.symbol.isStructuralRefinementMember case _ => false }
-        case x => throw new MatchError(x)
+
       }
       sym.name == nme.apply &&
-        !sym.hasStableFlag && // ???
-        sym.isCase &&
-        isClassTypeAccessible(tree) &&
-        !tree.tpe.finalResultType.typeSymbol.primaryConstructor.isLessAccessibleThan(tree.symbol)
+        sym.isCase && // only synthetic case apply methods
+        isClassTypeAccessible &&
+        !fun.tpe.finalResultType.typeSymbol.primaryConstructor.isLessAccessibleThan(sym)
     }
 
-    private def transformCaseApply(tree: Tree) = {
-      def loop(t: Tree): Unit = t match {
```

Comment:
no longer needed, handled by the recursive call to refcheck `transform(fun)` in `transformApplication`

Reply:
when I was looking at your branch locally yesterday, I didn't like this extra looping; I read the retronym ticket on "simple" case apply; and I already had a tweak commit to start refactoring, so I'm very glad you pushed on this!

I will fetch the latest to read instead of doing wordle over coffee.

Reply:
I will need a second cup of coffee after all.